improv_ble/improv_ble_client.py

```diff
@@ -0,0 +1,471 @@
+    CHARACTERISTIC_UUID_ERROR,
+    CHARACTERISTIC_UUID_RPC_COMMAND,
+    CHARACTERISTIC_UUID_RPC_RESULT,
+    CHARACTERISTIC_UUID_STATE,
+    SERVICE_UUID,
+    STATE_MAP,
+    parse_result,
+)
+from .util import try_parse_enum
+
+_LOGGER = logging.getLogger(__name__)
+
+_T = TypeVar("_T")
+_EnumT = TypeVar("_EnumT", bound=Enum)
+
+DISCONNECT_DELAY = 90
+
+DEFAULT_ATTEMPTS = 3
+
+
+def device_filter(advertisement_data: AdvertisementData) -> bool:
+    """Return True if the device is supported."""
+    uuids = advertisement_data.service_uuids
+    if SERVICE_UUID in uuids:
+        _LOGGER.error("Found device with service %s. ", SERVICE_UUID)
+        return True
+    return False
+
+
+class NotificationHandler:
+    """Container for notification handlers."""
+
+    error_callbacks: list[Callable[[prot.Error], None]]
+    state_callbacks: list[Callable[[prot.State], None]]
+
+    def __init__(self) -> None:
+        """Initialize."""
+        self.reset()
+
+    def notify(self, state: prot.Error | prot.State) -> None:
+        """Handle notification update."""
+        if isinstance(state, prot.Error):
+            for error_callback in self.error_callbacks:
+                error_callback(state)
+        else:
+            for state_callback in self.state_callbacks:
+                state_callback(state)
+
+    def reset(self) -> None:
+        """Reset."""
+        self.error_callbacks = []
+        self.state_callbacks = []
+
+    def subscribe_error(
+        self, callback: Callable[[prot.Error], None]
+    ) -> Callable[[], None]:
+        """Subscribe to error notifications."""
+
+        def remove() -> None:
+            self.error_callbacks.remove(callback)
+
+        self.error_callbacks.append(callback)
+        return remove
+
+    def subscribe_state(
+        self, callback: Callable[[prot.State], None]
+    ) -> Callable[[], None]:
+        """Subscribe to state notifications."""
+
+        def remove() -> None:
+            self.state_callbacks.remove(callback)
+
+        self.state_callbacks.append(callback)
+        return remove
+
+
+class ImprovBLEClient:
+    """Provision a device with support for Improv over BLE."""
+
+    _key_holder_id: bytes | None = None
+    _secret: bytes | None = None
+
+    def __init__(
+        self, ble_device: BLEDevice, advertisement_data: AdvertisementData | None = None
+    ):
+        """Initialize."""
+        self._advertisement_data = advertisement_data
+        self._ble_device = ble_device
+        self._client: BleakClient | None = None
+        self._notification_handlers = NotificationHandler()
+        self._response_handlers: dict[int, asyncio.Future[prot.Command]] = {}
+        self._connect_lock: asyncio.Lock = asyncio.Lock()
+        self._disconnect_reason: DisconnectReason | None = None
+        self._disconnect_timer: asyncio.TimerHandle | None = None
+        self._expected_disconnect: bool = False
+        self._procedure_lock: asyncio.Lock = asyncio.Lock()
```

Comment:
Is `asyncio.Lock` generic?

Reply:
I'm not sure I understand.

Do you mean to ask if `asyncio.Lock` is generic class? If so, the answer is that it isn't since an `asyncio.Lock` is just a stupid mutex which does not protect a certain resource or anything like that.